csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
 
 c10::intrusive_ptr<c10d::Work> Broadcast::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
 
   if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.dst_bufs.size() == 1) {
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
+    if (params_.is_root_in_mesh) {
+      // Do a local copy and the subsequent broadcast will be in place.
+      doLocalCopy(output_tensor, input_tensor);
     } else {
-      assertBufferCount(params_.dst_bufs, 0);
+      // `output_tensor` isn't allocated for this device.
+      output_tensor = input_tensor;
     }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
   }
 
   if (params_.team.size() == 1) {
     return nullptr;
   }
 
+  std::vector<at::Tensor> tensors({output_tensor});
   return comm.getBackendForTeam(params_.team, backend)
-      ->broadcast(
-          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
-          {.rootRank = root_relative_index_});
+      ->broadcast(tensors, {.rootRank = root_relative_index_});
 }
 
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
-}
+Gather::Gather(CommParams params) : Communication(params, "gather") {}
 
 c10::intrusive_ptr<c10d::Work> Gather::post(
```

Comment:
I'm confused. Something has to call PG with concrete `at::Tensor`s at run time. Who will be that caller if not Collective::post?

Reply:
My point is, in terms of interface and code organization purely, the function `post` shouldn't be member of the class `Collective` imo, because in general we want to separate symbolic IRs and runtime execution.
That's mostly incidental and not so important. Also, this is already what I did in this other pr https://github.com/NVIDIA/Fuser/pull/2185. So let me resolve the conversation

Reply:
Got it!